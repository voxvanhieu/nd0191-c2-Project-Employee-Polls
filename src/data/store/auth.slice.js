import { createSlice, createAsyncThunk, createReducer } from "@reduxjs/toolkit";

import * as API from '../api';
import { alertActions } from "./alert.slice";
import { history } from '../../helpers'

// Create slice
const name = 'auth';
const initialState = createInitialState();
const selectors = createSelectors();
const reducers = createReducers();
const extraActions = createExtraActions();
const slice = createSlice({ name, initialState, reducers, selectors });

// Exports
export const authActions = { ...slice.actions, ...extraActions };
export const authSelectors = { ...slice.selectors }
export const authReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        value: JSON.parse(localStorage.getItem('auth'))
    }
}

function createSelectors() {
    return {
        selectValue: (sliceState) => sliceState.value,
    }
}

function createReducers() {

    return {
        setAuth
    }

    function setAuth(state, action) {
        state.value = action.payload;
    }
}

function createExtraActions() {

    return {
        login: _login(),
        logout: _logout()
    }

    function _login() {
        return createAsyncThunk(
            `${name}/login`,
            async ({ username, password }, { dispatch }) => {
                // dispatch(alertActions.clear());
                try {
                    let user = await API.authenticateUser({ username, password });

                    dispatch(authActions.setAuth(user));

                    localStorage.setItem('auth', JSON.stringify(user));

                    if (user && user.name) {
                        dispatch(alertActions.success("Welcome"));
                        setTimeout(() => {
                            const { from } = history.location.state || { from: { pathname: '/' } };
                            history.navigate(from);
                        }, 500);
                    } else {
                        dispatch(alertActions.error("Wrong username or password"));
                    }
                } catch (error) {
                    console.log(error);
                }
            }
        )
    }

    function _logout() {
        return createAsyncThunk(
            `${name}/logout`,
            function (arg, { dispatch }) {
                dispatch(authActions.setAuth(null));
                localStorage.removeItem('auth');
                history.navigate('/login');
            }
        )
    }
}