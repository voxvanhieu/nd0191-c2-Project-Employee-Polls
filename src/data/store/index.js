import { configureStore } from '@reduxjs/toolkit';

import { alertReducer } from './alert.slice';
import { authReducer } from './auth.slice'
import { userReducer as usersReducer } from './user.slice'
import { pollsReducer } from './poll.slice';

export * from './alert.slice'
export * from './auth.slice'
export * from './user.slice'
export * from './poll.slice'

export const store = configureStore({
    reducer: {
        alert: alertReducer,
        auth: authReducer,
        users: usersReducer,
        polls: pollsReducer,
    }
})