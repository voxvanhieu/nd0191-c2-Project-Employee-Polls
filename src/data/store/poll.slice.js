import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import * as API from '../api';
import { useDispatch } from "react-redux";

// Create slice
const name = 'polls';
const initialState = createInitialState();
const selectors = createSelectors();
const advSelectors = createAdvancedSelectors();
const reducers = createReducers();
const extraActions = createExtraActions();
const extraReducers = createExtraReducers();

const slice = createSlice({ name, initialState, reducers, extraReducers, selectors });

// Exports
export const pollActions = { ...slice.actions, ...extraActions };
export const pollSelectors = { ...slice.selectors, ...advSelectors }
export const pollsReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        allIds: [],
        byId: {},
        status: {}
    }
}

function createSelectors() {
    return {
        selectValue: (sliceState) => sliceState.byId,
    }
}

//  Selector with parameters
function createAdvancedSelectors() {
    return {
        selectQuestionById: (pollId) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.includes(pollId)
                ? sliceState.byId[pollId]
                : null;
        },
        selectQuestionIds: (pollIds) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.filter(key => pollIds.includes(key));
        },
        selectQuestionIdsExcepts: (pollIds) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.filter(key => !pollIds.includes(key));
        }
    }
}


function createReducers() {
    return {
        getById: {
            reducer: (state, action) => state.byId[action.payload.id],
            prepare: (id) => ({ payload: id })
        },
    };
}

function createExtraActions() {

    return {
        getQuestions: createAsyncThunk(
            `${name}/getQuestions`,
            async () => await API._getQuestions()
        ),
    }
}

function createExtraReducers() {
    return (builder) => {
        buildReducerGetQuestions();

        function buildReducerGetQuestions() {
            var { pending, fulfilled, rejected } = extraActions.getQuestions;
            builder
                .addCase(pending, (state) => {
                    state.status = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.byId = action.payload;
                    state.allIds = Object.keys(state.byId);
                    state.status = { loading: false };
                })
                .addCase(rejected, (state, action) => {
                    state.status = { error: action.error };
                })
        }
    }
}
