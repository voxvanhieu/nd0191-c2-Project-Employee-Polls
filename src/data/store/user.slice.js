import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import * as API from '../api';

// Create slice
const name = 'users';
const initialState = createInitialState();
const selectors = createSelectors();
const advSelectors = createAdvancedSelectors();
const extraActions = createExtraActions();
const extraReducers = createExtraReducers();
const slice = createSlice({ name, initialState, extraReducers, selectors });

// Exports
export const userActions = { ...slice.actions, ...extraActions };
export const userSelectors = { ...slice.selectors, ...advSelectors };
export const userReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        allIds: [],
        byId: {},
        status: {}
    }
}

function createSelectors() {
    return {
        selectValue: (sliceState) => sliceState.byId,
    }
}

//  Selector with parameters
function createAdvancedSelectors() {
    return {
        selectUserInfoById: (userId) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.includes(userId) ? sliceState.byId[userId] : null;
        },
        selectCompletedPolls: (userId) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.includes(userId) ? sliceState.byId[userId].answers : null;
        }
    }
}

function createExtraActions() {

    return {
        getUsers: createAsyncThunk(
            `${name}/getUsers`,
            async () => await API._getUsers()
        )
    }
}

function createExtraReducers() {
    return (builder) => {
        buildReducerGetUsersAsync();

        function buildReducerGetUsersAsync() {
            var { pending, fulfilled, rejected } = extraActions.getUsers;
            builder
                .addCase(pending, (state) => {
                    state.status = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.byId = action.payload;
                    state.allIds = Object.keys(action.payload);
                    state.status = { loading: false };
                })
                .addCase(rejected, (state, action) => {
                    state.status = {
                        loading: false,
                        error: action.error
                    };
                })
        }
    }
}