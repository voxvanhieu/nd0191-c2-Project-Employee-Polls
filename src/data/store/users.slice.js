import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import * as API from '../api';

// Create slice
const name = 'users';
const initialState = createInitialState();
const extraActions = createExtraActions();
const extraReducers = createExtraReducers();
const slice = createSlice({ name, initialState, extraReducers });

// Exports
export const userAction = { ...slice.actions, ...extraActions };
export const usersReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        list: null,
        item: null
    }
}

function createExtraActions() {

    return {
        register: _register(),
        getAll: _getAll(),
        getByUsername: _getByUsername(),
    }

    function _register() {
        // TBD
        return createAsyncThunk(
            `${name}/register`,
            async (user) => await new Promise((resolve) => {
                setTimeout(() => {
                    resolve(user);
                }, 1000);
            })
        );
    }

    function _getAll() {
        return createAsyncThunk(
            `${name}/getAll`,
            async () => await API._getUsers()
        );
    }

    function _getByUsername() {
        return createAsyncThunk(
            `${name}/getById`,
            async (id) => await API._getUserByUsername(id)
        );
    }
}

function createExtraReducers() {
    return (builder) => {
        buildReducerGetAll();
        buildReducerGetByUsername();

        function buildReducerGetAll() {
            var { pending, fulfilled, rejected } = extraActions.getAll;
            builder
                .addCase(pending, (state) => {
                    state.list = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.list = { value: action.payload };
                })
                .addCase(rejected, (state, action) => {
                    state.list = { error: action.error };
                })
        }

        function buildReducerGetByUsername() {
            var { pending, fulfilled, rejected } = extraActions.getByUsername;
            builder
                .addCase(pending, (state) => {
                    state.item = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.item = { value: action.payload };
                })
                .addCase(rejected, (state, action) => {
                    state.item = { error: action.error };
                })
        }
    }
}