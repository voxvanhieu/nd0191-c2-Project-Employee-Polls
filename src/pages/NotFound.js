import React from 'react';
import { useNavigate } from 'react-router-dom';

const NotFound = () => {
    const navigate = useNavigate();

    const handleBackToHome = () => {
        navigate('/');
    };

    return (
        <div className="flex flex-col items-center justify-center min-h-screen text-center bg-gray-100">
            <h1 className="text-6xl font-bold text-primary-600">404</h1>
            <p className="mt-4 text-2xl text-black">Oops! The page you're looking for doesn't exist.</p>
            <button
                onClick={handleBackToHome}
                className="px-6 py-3 mt-6 text-lg font-medium text-white bg-blue-600 rounded hover:bg-blue-700"
            >
                Back to Home
            </button>
        </div>
    );
};

export { NotFound };
