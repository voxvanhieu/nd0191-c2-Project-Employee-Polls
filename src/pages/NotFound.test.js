import React from 'react';
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { render } from '@testing-library/react';
import { store } from '../data/store';
import { NotFound } from './NotFound';

describe('App snapshot test', () => {
  it('will match snapshot', () => {
    let component = render(
      <React.StrictMode>
        <Provider store={store}>
          <BrowserRouter>
            <NotFound />
          </BrowserRouter>
        </Provider>
      </React.StrictMode>
    );
    expect(component).toMatchSnapshot();
  })
})
