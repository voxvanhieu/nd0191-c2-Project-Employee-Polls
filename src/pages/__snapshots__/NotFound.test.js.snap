// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot test will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col items-center justify-center min-h-screen text-center bg-gray-100"
      >
        <h1
          class="text-6xl font-bold text-primary-600"
        >
          404
        </h1>
        <p
          class="mt-4 text-2xl text-black"
        >
          Oops! The page you're looking for doesn't exist.
        </p>
        <button
          class="px-6 py-3 mt-6 text-lg font-medium text-white bg-blue-600 rounded hover:bg-blue-700"
        >
          Back to Home
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col items-center justify-center min-h-screen text-center bg-gray-100"
    >
      <h1
        class="text-6xl font-bold text-primary-600"
      >
        404
      </h1>
      <p
        class="mt-4 text-2xl text-black"
      >
        Oops! The page you're looking for doesn't exist.
      </p>
      <button
        class="px-6 py-3 mt-6 text-lg font-medium text-white bg-blue-600 rounded hover:bg-blue-700"
      >
        Back to Home
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
